File: 20231007_Saturday/20231007.py
# https://leetcode.com/problems/build-array-where-you-can-find-the-maximum-exactly-k-comparisons/solution/
class Solution:
    def numOfArrays(self, n: int, m: int, k: int) -> int:    
        dp = [[[0] * (k + 1) for _ in range(m + 1)] for __ in range(n + 1)]
        MOD = 10 ** 9 + 7
        
        for num in range(len(dp[0])):
            dp[n][num][0] = 1
        
        for i in range(n - 1, -1, -1):
            for max_so_far in range(m, -1, -1):
                for remain in range(k + 1):
                    ans = (max_so_far * dp[i + 1][max_so_far][remain]) % MOD
    
                    if remain > 0:
                        for num in range(max_so_far + 1, m + 1):
                            ans = (ans + dp[i + 1][num][remain - 1]) % MOD
                        
                    dp[i][max_so_far][remain] = ans

        return dp[0][0][k]

File: 20231007_Saturday/20231007.go
// Leetcode: 1420. Build Array Where You Can Find The Maximum Exactly K Comparisons.
// https://leetcode.com/problems/build-array-where-you-can-find-the-maximum-exactly-k-comparisons/
package main

func numOfArrays(n int, m int, k int) int {
	mod := 1_000_000_007
	dp := make([][][]int,n+1)
	for i := 0; i <= n ; i += 1 {
		dp[i] := make([][]int,m+1)
		for j := 0; j <= m; j += 1 {
			dp[i][j] = make([]int, k+1)
		}
	}
	for j := 0; j <= m; j += 1 {
		dp[n][j][0] = 1
	}
	return dp[0][0][k]
}
